### PR TITLE
pm: fix issue that system crash when passed invalid relpath value

### PR DESCRIPTION
**Pull request by gneworld**

### CODE REVIEW

File: drivers/power/pm/pm_procfs.c [hunk 1]
```diff
@@ -159,6 +161,23 @@ static FAR const char *g_pm_state[PM_COUNT] =
  * Private Functions
  ****************************************************************************/
```
Comment (anchao) [inline drivers/power/pm/pm_procfs.c:163]:
```
/****************************************************************************
 * Name: pm_path_validate
 ****************************************************************************/
```

File: drivers/power/pm/pm_procfs.c [hunk 2]
```diff
@@ -159,6 +161,23 @@ static FAR const char *g_pm_state[PM_COUNT] =
...
 
+static int pm_path_validate(FAR const char *relpath)
+{
+  int index = -1;
+
+  for (int i = 0; i < nitems(g_pm_files); i++)
```
Comment (anchao) [inline drivers/power/pm/pm_procfs.c:168]: move `int i` out of loop

File: drivers/power/pm/pm_procfs.c [hunk 3]
```diff
@@ -191,6 +210,11 @@ static int pm_open(FAR struct file *filep, FAR const char *relpath,
     }
 
   relpath += strlen("pm/");
+  if (pm_path_validate(relpath) < 0)
+    {
```
Comment (anchao) [inline drivers/power/pm/pm_procfs.c:214]: memory leak

File: drivers/power/pm/pm_procfs.c (hunk 1 above)
Comment (gneworld) [reply drivers/power/pm/pm_procfs.c:163]: done

File: drivers/power/pm/pm_procfs.c (hunk 2 above)
Comment (gneworld) [reply drivers/power/pm/pm_procfs.c]: done

File: drivers/power/pm/pm_procfs.c (hunk 3 above)
Comment (gneworld) [reply drivers/power/pm/pm_procfs.c]: done

File: drivers/power/pm/pm_procfs.c [hunk 4]
```diff
@@ -181,6 +203,12 @@ static int pm_open(FAR struct file *filep, FAR const char *relpath,
       return -EACCES;
     }
 
+  relpath += strlen("pm/");
+  if (pm_path_validate(relpath) < 0)
```
Comment (xiaoxiang781216) [inline drivers/power/pm/pm_procfs.c:207]:
```suggestion
  i = pm_path_validate(relpath);
  if (i < 0)
```
and change line 221 to 231:
```
  pmfile->read = g_pm_files[i].read;
  pmfile->domain = atoi(relpath + strlen(g_pm_files[i].name));
```

File: drivers/power/pm/pm_procfs.c [hunk 5]
```diff
@@ -114,6 +114,8 @@ static int     pm_rewinddir(FAR struct fs_dirent_s *dir);
 
 static int     pm_stat(FAR const char *relpath, FAR struct stat *buf);
 
+static int     pm_path_validate(FAR const char *relpath);
```
Comment (xiaoxiang781216) [inline drivers/power/pm/pm_procfs.c:117]:
```suggestion
static int     pm_get_file_index(FAR const char *relpath);
```

Comment (gneworld) [reply drivers/power/pm/pm_procfs.c]: done

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**